### PR TITLE
sequence_mask

### LINKED ISSUE #20981: sequence_mask

### PR DESCRIPTION
**Pull request by usayrimrehman**
- Close #20981

### CODE REVIEW

**Comment by usayrimrehman** [comment]
@kurshakuz
Hey,
Can you please give the PR a look, and lemme know if you need any changes.

**Review by kurshakuz — CHANGES_REQUESTED** [review]
Hi! Seems like there is a resulting dtype mismatch that produces the following error:
```
E           AssertionError: Cannot cast from int64 to int32
```
Can you please resolve it and test it out? Thanks!

**Comment by usayrimrehman** [comment]
@kurshakuz 
I changed the test a little because the values in lengths are always <= maxlen, so the casting problem is a non-issue since now both of them will have the same dtype. The problem now is that the tests keep getting terminated without running completely, and I'm not sure why that might be. I'd appreciate if you could give it a look, and lemme know if it's a problem on my side.
image

**Comment by usayrimrehman** [comment]
@kurshakuz 
Hey, I looked into the issue that you mentioned, and fixed the test accordingly. Since all values in `lengths` <= `maxlen`, the casting problem is a non-issue. You can also see TensorFlow's comment on the issue below.

image

I'd appreciate it if you give the PR another look, and lemme know if it's good to go.

**Comment by usayrimrehman** [comment]
@KareemMAX 
Hey, can you please give this PR a look, and merge it, so I can move on to the next task because the assigned reviewer appears to be busy.
Thank you in advance.

File: ivy/functional/frontends/tensorflow/general_functions.py [hunk 1]
```diff
@@ -650,3 +651,21 @@ def truncatemod(x, y):
...
+        maxlen = ivy.maximum(0, maxlen)
+    else:
+        maxlen = ivy.array(maxlen)
+    if ivy.get_num_dims(maxlen )is not None and ivy.get_num_dims(maxlen) != 0:
+        raise ValueError("Argument `maxlen` must be scalar for sequence_mask, "
+                         f"received `maxlen` = {maxlen} "
+                         f"with shape '{maxlen.get_shape()}' instead")
+
+    row_vector = ivy.arange(0, int(maxlen), 1)
+    matrix = cast(ivy.expand_dims(lengths, axis=-1), maxlen.dtype)._ivy_array
```
Comment (KareemMAX) [inline ivy/functional/frontends/tensorflow/general_functions.py:669]: Are you sure that using `_ivy_array` is a good practice?

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py [hunk 2]
```diff
@@ -2329,3 +2329,37 @@ def test_tensorflow_unravel_index(
...
+# sequence_mask
+@handle_frontend_test(
+    fn_tree="tensorflow.sequence_mask",
+    dtype_and_lens=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("integer"),
+        num_arrays=1,
+        min_value=1,
+        max_value=5,
+    ),
+    max_len=st.integers(min_value=5, max_value=1000),
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py:2342]: I feel like this test is not exhaustive, maybe consider doing a compositional strategy instead. As I fear hard-coded values like this might not discover hidden cases.
Also don't forget to check different dtypes

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py [hunk 3]
```diff
@@ -2329,3 +2329,37 @@ def test_tensorflow_unravel_index(
...
+    max_len,
+    frontend,
+    test_flags,
+    fn_tree,
+    on_device,
+    backend_fw,
+):
+    input_dtype, lens = dtype_and_lens
+    helpers.test_frontend_function(
+        input_dtypes=['int32'],
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py:2357]: I think this is not correct, shouldn't it be `input_dtype`?

**Review by KareemMAX — CHANGES_REQUESTED** [review]
Hey @usayrimrehman,

I've pointed out a few comments, also it seems like you didn't setup linting correctly. I'll run `ivy-gardener` at the meantime put would be better if you setup `pre-commit` by following [our guide](https://unify.ai/docs/ivy/overview/contributing/setting_up.html#pre-commit).

Also, do you mind making a comment with "add_frontend_checklist" without the double quotes nor new lines or extra spaces. Once you make that comment a checklist will appear, where you can check which things you feel is completed in your implementation

File: ivy/functional/frontends/tensorflow/general_functions.py (hunk 1 above)
Comment (usayrimrehman) [reply ivy/functional/frontends/tensorflow/general_functions.py:669]: The `cast` function outputs an eager tensor, and it's not possible to perform the less than operation on that in the next line. It throws an error. So I convert the `matrix` into an Ivy Array which makes it possible to find the result later.

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py (hunk 3 above)
Comment (usayrimrehman) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py:2357]: The input_dtype comes out to be int64. This throws an error when I later attempt to cast it to maxlen's dtype saying it's not possible to cast from int64 to int32. This is not an issue at all as every value in the input array is less than maxlen, so if maxlen fits into int32, then so should the input array.

**Comment by usayrimrehman** [comment]
Hey @KareemMAX,
I've replied to your comments. Lemme know if they make, or if other subsequent changes are needed.

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py (hunk 2 above)
Comment (usayrimrehman) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py:2342]: Plus, both the input and maxlen need to be integers, which is why I didn't explore other dtypes

Comment (KareemMAX) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py:2342]: I would suggest giving them a chance with using compositional strategy, unit tests should be designed to discover those edge cases. So they don't haunt us in the future

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py (hunk 3 above)
Comment (KareemMAX) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py:2357]: That's a fault with your value generation then. `helpers.get_dtypes("integer")` generate both `int32` and `int64`, you are trying to convince the test function that you are sending a dtype different from what you are actually sending. Which is not right

**Review by KareemMAX — CHANGES_REQUESTED** [review]
Have made some comments + asked internally about the `cast` issue. Please fix the linting and address the comments in the mean time

File: ivy/functional/frontends/tensorflow/general_functions.py (hunk 1 above)
Comment (KareemMAX) [reply ivy/functional/frontends/tensorflow/general_functions.py:669]: I guess it would be better if you used:
```python
ivy.astype(ivy.expand_dims(lengths, axis=-1), maxlen.dtype)
OR
ivy.expand_dims(lengths, axis=-1).astype(maxlen.dtype)
```
Does those code fragments have issues?

Comment (usayrimrehman) [reply ivy/functional/frontends/tensorflow/general_functions.py]: I removed the cast function as a whole, as it was just a precautionary measure.

**Comment by usayrimrehman** [comment]
@KareemMAX 
Have incorporated most of the changes you requested apart from the fact that I have to assign a maximum value to `maxlen` otherwise the tests get terminated. You can try it for yourself, and let me know if it's a problem on my end.

**Comment by KareemMAX** [comment]
ivy-gardener

**Comment by KareemMAX** [comment]
add_frontend_checklist

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py [hunk 4]
```diff
@@ -2329,3 +2329,47 @@ def test_tensorflow_unravel_index(
...
+        helpers.dtype_and_values(
+            available_dtypes=helpers.get_dtypes("integer"),
+            num_arrays=1,
+            min_value=1,
+            max_value=max_val,
+        )
+    )
+
+    max_len = draw(st.integers(min_value=int(ivy.max(x)), max_value=max_val))
+    return dtype, x, max_len
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py:2348]: What is `x`? I guess it would be better if you changed it to `lens`

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py [hunk 5]
```diff
@@ -2329,3 +2329,47 @@ def test_tensorflow_unravel_index(
...
+
+# sequence_mask
+@st.composite
+def _sequence_mask_helper(draw):
+    max_val = draw(st.integers(min_value=1, max_value=100000))
+    dtype, x = draw(
+        helpers.dtype_and_values(
+            available_dtypes=helpers.get_dtypes("integer"),
+            num_arrays=1,
+            min_value=1,
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py:2342]: Does `lens` need to be positive?

**Comment by usayrimrehman** [comment]
@KareemMAX 
I've made the the requested changes.

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py [hunk 6]
```diff
@@ -2355,6 +2356,62 @@ def test_tensorflow_unravel_index(
...
+    'uint8',
+    'complex128',
+    'bool',
+    'float64',
+    'int8',
+    'int16',
+    'complex64',
+    'float32',
+    'int32',
+    'int64']))
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py:2384]: Why don't you use `helpers.get_dtypes("valid")`?
```suggestion
    dtype = draw(helpers.get_dtypes("valid"))
```

File: ivy/functional/frontends/tensorflow/general_functions.py [hunk 7]
```diff
@@ -657,6 +658,32 @@ def unravel_index(indices, dims, out=None, name=None):
...
+            f"with shape '{maxlen.get_shape()}' instead"
+        )
+
+    row_vector = ivy.arange(0, int(maxlen), 1)
+    matrix = ivy.expand_dims(lengths, axis=-1)
+    result = row_vector < matrix
+    if dtype == None:
+        return result
+    else:
+        return cast(result, dtype)
```
Comment (KareemMAX) [inline ivy/functional/frontends/tensorflow/general_functions.py:684]: Are you sure about the use of cast here? I believe it's best if you used `astype`

```suggestion
        return ivy.astype(result, dtype)
```

**Review by KareemMAX — CHANGES_REQUESTED** [review]
Hey @usayrimrehman,
I've made some other comments that I need you to fix before merging. Also I've noticed you forgot to setup `pre-commit` please use [this guide](https://unify.ai/docs/ivy/overview/contributing/setting_up.html#pre-commit) to help you out.

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py (hunk 6 above)
Comment (usayrimrehman) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py:2384]: 1) This draws a list of dtypes instead of a single dtype.
2) This ignores a few dtypes that I believe should be considered by the testing function

File: ivy/functional/frontends/tensorflow/general_functions.py [hunk 8]
```diff
@@ -9,6 +9,7 @@
     handle_tf_dtype,
     to_ivy_dtype,
 )
+from ivy.functional.frontends.tensorflow.dtypes import cast
```
Comment (KareemMAX) [inline ivy/functional/frontends/tensorflow/general_functions.py:12]: This is not needed I believe, can you remove it?

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py (hunk 6 above)
Comment (KareemMAX) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_general_functions.py:2384]: Makes sense, as it only allow you to draw from supported dtypes. Which are `lens` dtypes.

**Comment by usayrimrehman** [comment]
Hey @KareemMAX
I've removed all the unnecessary imports.

**Comment by KareemMAX** [comment]
ivy-gardener

**Review by KareemMAX — CHANGES_REQUESTED** [review]
Hey @usayrimrehman,
Great work, only issue we are having now is `lens` being positive. I think this constraint it not needed. Can you fix that?

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by usayrimrehman** [comment]
Still working on this

**Comment by usayrimrehman** [comment]
Hey @KareemMAX
I've made the changes to accommodate the negative values for lengths. Can you please check if it's good to go now.

**Comment by usayrimrehman** [comment]
ivy-gardener

**Comment by usayrimrehman** [comment]
ivy-gardener

**Review by KareemMAX — CHANGES_REQUESTED** [review]
I've unresolved an issue that it seems you didn't fix. Can you fix that? Also resolve the merge conflicts please

File: ivy/functional/frontends/tensorflow/general_functions.py [hunk 9]
```diff
@@ -409,6 +409,34 @@ def searchsorted(sorted_sequence, values, side="left", out_type="int32"):
...
+        raise ValueError(
+            "Argument `maxlen` must be scalar for sequence_mask, "
+            f"received `maxlen` = {maxlen} "
+            f"with shape '{maxlen.get_shape()}' instead"
+        )
+
+    row_vector = ivy.arange(0, int(maxlen), 1)
+    matrix = ivy.expand_dims(lengths, axis=-1)
+    result = row_vector < matrix
+    if dtype == None:
```
Comment (KareemMAX) [inline ivy/functional/frontends/tensorflow/general_functions.py:434]:
```suggestion
    if dtype is None:
```

**Review by KareemMAX — APPROVED** [review]
LGTM! I've made a small suggestion that I'll apply
